http_logs/challenges/common/update-querying-schedule.json

```diff
@@ -0,0 +1,155 @@
+        {
+          "operation": "delete-index"
+        },
+        {
+          "operation": {
+            "operation-type": "create-index",
+            "settings": {%- if index_settings is defined %} {{index_settings | tojson}} {%- else %} {
+              "index.number_of_shards": {{number_of_shards | default(1)}},
+              "index.number_of_replicas": {{number_of_replicas | default(1)}},
+              "index.store.type": "{{store_type | default('fs')}}"
+            }{%- endif %}
+          }
+        },
+        {
+          "name": "check-cluster-health",
+          "operation": {
+            "operation-type": "cluster-health",
+            "index": "logs-*",
+            "request-params": {
+              "wait_for_status": "{{cluster_health | default('green')}}",
+              "wait_for_no_relocating_shards": "true"
+            },
+            "retry-until-success": true
+          }
+        },
+        {
+          "operation": "update",
+          "warmup-time-period": 240,
+          "clients": {{bulk_indexing_clients | default(8)}},
```

Comment:
> Does this mean 8 clients, running at "full pace", i.e., as soon as one operation completes, the next will start?

Each line of a corpus file contains a single document (without IDs) to be indexed by Rally during benchmarking. When using 8 clients, Rally will split each corpus using as many splits as there are clients for bulk indexing parallelization. Each client performs full-pace bulk indexing at the same time, working on its assigned documents according to the offset table.

> It is not entirely clear how the conflict id generation works. I guess we can see in monitoring how many updates vs appends that were done? That would be interesting to know I think. 

A bulk indexing operation without ID conflicts relies on ID generation in Elasticearch. Inserting conflicts into the mix instructs Rally to do sequential ID generation over the entire corpus, factoring in the offset to keep each client indexing within its own range.

A Rally bulk update task with detailed reporting enabled captures these additional metrics as part of latency, service time, and processing time [^1]:

[^1]: https://esrally.readthedocs.io/en/latest/metrics.html?highlight=metrics#metric-keys

```json
"meta": {
      "distribution_version": "8.9.0-SNAPSHOT",
      "distribution_flavor": "default",
      "index": "logs-241998",
      "took": 532,
      "success": true,
      "success-count": 5000,
      "error-count": 0,
      "ops": {
        "index": {
          "item-count": 3692,
          "created": 3692
        },
        "update": {
          "item-count": 1308,
          "updated": 1308
        }
      },
      "shards_histogram": [
        {
          "item-count": 5000,
          "shards": {
            "total": 1,
            "successful": 1,
            "failed": 0
          }
        }
      ],
      "bulk-request-size-bytes": 966909,
      "total-document-size-bytes": 680601
    }
```

> As discussed elsewhere, ideally we'd have a set of documents already and then only do updates on those documents, making sure all requests are updates. We'd want a decent amount of actual updates (i.e., conflicts) in this benchmark since otherwise we may not be able to see any problems with updates.

`conflict-probability` could be set to 100% which should have the same effect.

Reply:
Would `conflict-probability` not result in very few documents in ES that were updated all the time? I.e., the indexed set of data would be tiny.

Reply:
You're right, @henningandersen, and I misinterpreted the use of the parameter. `conflict-probability: 100` indexes a single document per index client, then everything else is a conflict. Would using 50% or 75%, resulting in indexed documents of 50% & 25% respectively be enough to see problems with updates? http_logs contains **247,249,096** documents.

| conflict_probability | indexed documents |
| --- | --- |
| 25 | 185446537 (75%) |
| 50 | 123624548 (50%) |
| 75 | 61812274 (25%) |